Sources/SwiftFrameCore/Workers/ConfigProcessor.swift

```diff
@@ -27,50 +32,60 @@ public class ConfigProcessor {
     }
 
     public func run() throws {
-        if verbose {
+        if verbose && !noManualValidation {
             data.printSummary(insetByTabs: 0)
             print("Press return key to continue")
             _ = readLine()
         }
 
-        print("Parsed and validated config file")
+        print("Parsed and validated config file\n")
 
         if data.clearDirectories {
-            print("Clearing Output Directories")
+            let clearingStart = CFAbsoluteTimeGetCurrent()
             try FileManager.default.ky_clearDirectories(data.outputPaths, localeFolders: Array(data.titles.keys))
+            printElapsedTime("Clear output directories", startTime: clearingStart)
         }
 
-        print("Rendering...\n")
+        let stringProcessingStart = CFAbsoluteTimeGetCurrent()
+
+        try data.deviceData.forEach { deviceData in
+            try deviceData.screenshotsGroupedByLocale.forEach { locale, _ in
+                printVerbose("Processing strings for locale \(locale) (\(deviceData.outputSuffixes.first ?? "unknown device"))")
+
+                let associatedStrings = try data.makeAssociatedStrings(for: deviceData, locale: locale)
+                let fontSizesByGroup = try data.makeSharedFontSizes(for: associatedStrings)
+                try AttributedStringCache.shared.process(
+                    associatedStrings,
+                    locale: locale,
+                    deviceIdentifier: deviceData.outputSuffixes.joined(),
+                    maxFontSizeByGroup: fontSizesByGroup,
+                    font: try data.fontSource.font(),
+                    color: data.textColorSource.color,
+                    maxFontSize: data.maxFontSize
+                )
+            }
+        }
 
-        // We need a special semaphore here, since the creation of the attributed strings has to happen
-        // on the main thread and anything else can happen asynchronously but we still want to finish
-        // execution only when everything has finished
-        let semaphore = RunLoopSemaphore()
+        printElapsedTime("Parse HTML strings into attributed strings", startTime: stringProcessingStart)
 
-        let start = CFAbsoluteTimeGetCurrent()
+        print("\nRendering...\n")
 
-        DispatchQueue.global(qos: .userInitiated).async {
-            let group = DispatchGroup()
+        let imageRenderingStart = CFAbsoluteTimeGetCurrent()
 
-            self.data.deviceData.forEach { deviceData in
-                group.enter()
-                DispatchQueue.global(qos: .userInitiated).ky_asyncOrExit { [weak self] in
-                    try self?.process(deviceData: deviceData)
-                    group.leave()
-                }
-            }
+        let group = DispatchGroup()
 
-            group.wait()
-            DispatchQueue.main.async {
-                semaphore.signal()
+        data.deviceData.forEach { deviceData in
```

Comment:
**question** now that we don't need the `RunloopSemaphore`, we could use `DispatchQueue.concurentPerform` which should also remove the need for a dispatch group.

Reply:
Ah you're right, yeah I tried it out before extracting the string parsing and that didn't work. Seems to be faster with `DispatchQueue.concurentPerform` too!